src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php

```diff
@@ -290,6 +290,42 @@ protected function stream(string $view, array $parameters = [], StreamedResponse
         return $response;
     }
 
+    /**
+     * Handles a form:
+     *
+     * * if the form is not submitted, the Twig template passed in $view is rendered and a 200 HTTP status code is set
+     * * if the form is submitted but invalid, the Twig template passed in $view is rendered and 422 HTTP status code is set
+     * * if the form is submitted and valid, the entity is saved (only if it is managed by Doctrine ORM), a 306 HTTP status code is set and the Location HTTP header is set to the value of $redirectUrl
+     */
+    public function handleForm(Request $request, FormInterface $form, string $view, ?string $redirectUrl = null, ?callable $onSuccess = null, array $viewParameters = [], ?Response $response = null): Response
+    {
+        $form->handleRequest($request);
+
+        $submitted = $form->isSubmitted();
+        $valid = $submitted && $form->isValid();
+
+        if ($valid) {
+            if (null !== $onSuccess) {
+                return $onSuccess();
+            }
+
+            $data = $form->getData();
+            if ($manager = $this->getDoctrine()->getManagerForClass(get_class($data))) {
+                $manager->persist($data);
+                $manager->flush();
+            }
+
+            return $this->redirect($redirectUrl, Response::HTTP_SEE_OTHER);
+        }
+
+        $response = $this->render($view, ['form' => $form->createView()] + $viewParameters, $response);
```

Comment:
To me, this line should be removed (alongside with the corresponding arguments).
The current design forces populating `$viewParameters` even if it's never used (when the form is successful.)

That'd mean that the consumer code should add more lines after calling this helper, eg:

```php
$form = $this->createForm($formType, $data, $formOptions);
$response = $this->handleForm($request, $form, function () {
    // ...
    return $this->redirect(...);
});
if ($response->isRedirect()) {
    return $response;
}

// [...] compute view parameters

return $this->render('...', ['form' => $form->createView(), ...], $response);
```

That's a bit more verbose I agree...

Reply:
In order to remove the `if ($response->isRedirect()) {` block, we could throw a signaling exception in handleForm when the closure returns a redirect.

I think that would be my favorite design.

Reply:
Maybe better - we would pass two closures to `handleForm()`:

```php
$form = $this->createForm($formType, $data, $formOptions);

return $this->handleForm($request, $form, function () {
    // ...
    return $this->redirect(...);
}, function () {
    // [...] compute view parameters

    return $this->render('...', ['form' => $form->createView(), ...]);
}
```

Reply:
I'd be :+1: with this:
```php
public function handleForm(FormInterface $form, Request $request, callable $onSuccess, callable $render): Response
{
    $form->handleRequest($request);
    $submitted = $form->isSubmitted();
    $valid = $submitted && $form->isValid();
    if ($valid) {
        return $onSuccess($form, $request);
    }
    $response = $render($form, $request);
    if ($submitted && $response->isSuccessful()) {
        $response->setStatusCode(Response::HTTP_UNPROCESSABLE_ENTITY);
    }
    return $response;
}
```